### PR TITLE
Generalized callbacks / app flow

### PR DESCRIPTION
**Pull request by wpears**
Includes:
- error first callbacks (so errors aren't just swallowed by the api)
- updates to IRS and signature to use standard callback patterns
- expands and refactors the API
- allows transitioning to IRS report after verifying all quality/macro

Going to integrate this with the q029/q595 work tomorrow; but this kinda ballooned as I realized more stuff needed to be changed, so I wanted it in ASAP

### CODE REVIEW

File: __tests__/Signature.js [hunk 1]
```diff
@@ -5,34 +5,34 @@ var ReactDOM = require('react-dom');
...
-    receipt: 'somehash'
+    receipt: 'dc9e5827abb678f54103e7b89435abf9b36648797ebb6516a52ab33ab4e46cee'
   });
+  code = code === 13 ? 12 : 13;
 });
 
 describe('irs report', function(){
-  var changeHandlerTrue = function(e){
-    expect(e.target.checked).toBeTruthy();
+  var changeHandlerTrue = function(err, status){
```
Comment (awolfe76) [inline __tests__/Signature.js]: I like the way you did things in the IRS test, for [example](https://github.com/cfpb/hmda-platform-ui/pull/142/files#diff-b28d2400315fa2e86b0451fc245b96bbR56). Both the function names and declaring them inside the `it()`.
Maybe do that here too?

**Comment by awolfe76** [comment]
**There's a conflict: "This branch has conflicts that must be resolved".**

_The rest of this may be coming but wanted to point it out._

Still digging through it but when looking at the UI:
- for `Bank 0` after selecting a file and clicking upload no progress happens
- for `Bank 3`, it's "in progress" but when viewing the progress you see the file progress completes but no edits appear
- for `Bank 5`, its "complete" but there is no receipt and timestamp rendered (the checkbox is checked)
- **You mention is above** but I don't see the Q029 or Q595 edits on any of the banks
- Verification of a quality or macro edit removes the IRS report and signature

> allows transitioning to IRS report after verifying all quality/macro

I can't test this in the UI because several don't render the edits. _Seems_ to only render correctly when `code > 9`. I don't see edits unless the IRS report is there already.

**Comment by wpears** [comment]
Currently ignoring q029/q595 wrt app state/flow.. since I think they should be rolled in with other quality edits 😎  (see #144)

**Comment by awolfe76** [comment]
Looks like the latest commits fixed all but 

> Verification of a quality or macro edit removes the IRS report and signature

After entering text for a macro edit, clicking verify removes the IRS and signature components. The others are 👍  though.

**Comment by wpears** [comment]
> Verification of a quality or macro edit removes the IRS report and signature

Which bank are you testing on? This is somewhat hacky because, in truth, IRS report and signature would never show if macro/quality haven't been verified

I'm guessing you're seeing this behavior because the mock data is in an impossible state.

That said, _unchecking_ a verified macro/quality _should_ remove IRS and signature components, no? Because they only show when all macro/quality are verified.

**Comment by awolfe76** [comment]
I'm using `/bank6id/2017/2`.

You are right about 

> That said, unchecking a verified macro/quality should remove IRS and signature components, no? Because they only show when all macro/quality are verified.

But this is happening when I check a verification, which should do the opposite ... if all others are checked to. But it shouldn't remove those components.

**Comment by wpears** [comment]
Right; but there is no ways those (IRS and signature) should be shown if everything isn't already validated. They only are now because we're using one set of edits for all the institutions. I can add additional json documents and the logic to route to the correct one, if you'd prefer. That'd probably be in a different PR, though

**Comment by awolfe76** [comment]
Got it. I see what's going on now. Sorry for the confusion. Merging.

**Comment by awolfe76** [comment]
And I think we're ok, no need to add more json. I think we'll get better once the back-end API stuff gets going.

**Pull request merged**